### PR TITLE
HEEDLS-603 - set up supervisor delegate relation when bulk uploading …

### PR DESCRIPTION
**Pull request by DanBloxham-sw**
…delegates

### JIRA link
https://softwiretech.atlassian.net/browse/HEEDLS-603

### Description
Added some calls to the supervisor delegate service during the bulk upload process for registrations.

### Screenshots
N/A

-----
### Developer checks
(Leave tasks unticked if they haven't been appropriate for your ticket.)

I have:
- [x] Run the formatter and made sure there are no IDE errors.
- [x] Written tests for the changes (controller, data services, services, view models etc) and manually tested my work with and without JavaScript. Full manual testing guidelines can be found here: https://softwiretech.atlassian.net/wiki/spaces/HEE/pages/6703648740/Testing
- [ ] Updated/added documentation in Swiki and/or Readme.
- [ ] Updated my Jira ticket with information about other parts of the system that were touched as part of the MR and have to be sanity tested to ensure nothing’s broken.
- [x] Scanned over my own MR to ensure everything is as expected.

### CODE REVIEW

File: DigitalLearningSolutions.Data/Services/SupervisorDelegateService.cs
Comment (DanBloxham-sw) [inline DigitalLearningSolutions.Data/Services/SupervisorDelegateService.cs:27]: I saw that this userDataService wasn't being used so I removed it.

File: DigitalLearningSolutions.Data.Tests/Services/DelegateUploadFileServiceTests.cs
Comment (livzorn) [inline DigitalLearningSolutions.Data.Tests/Services/DelegateUploadFileServiceTests.cs:642]: Confused by the variable name a bit, does RegisterDelegateByCentre actually return a status code? I see the return variable in the method is called candidateNumber

**Review by livzorn — COMMENTED** [review]
Looks good, just one question

File: DigitalLearningSolutions.Data.Tests/Services/DelegateUploadFileServiceTests.cs
Comment (DanBloxham-sw) [reply DigitalLearningSolutions.Data.Tests/Services/DelegateUploadFileServiceTests.cs:642]: RegisterDelegateByCentre calls a stored procedure and returns what it returns. This can either by a candidate number if a delegate is registered successfully, or a non-zero status code if something goes wrong. These tests are specifically testing those error status codes prevent further processing.

Comment (livzorn) [reply DigitalLearningSolutions.Data.Tests/Services/DelegateUploadFileServiceTests.cs:642]: Understood, thanks for the clarification! :)

**Review by livzorn — APPROVED** [review]

File: DigitalLearningSolutions.Data/Services/DelegateUploadFileService.cs
Comment (SteveJacksonSoft) [inline DigitalLearningSolutions.Data/Services/DelegateUploadFileService.cs:224]: I think this would be better placed outside this method.
We've no guarantee that this method will only be called just after the delegate has successfully been registered, so can't safely use the null assertion here. But if we move it outside the method, into the switch statement, we can.
Also the method is kind of doing two things currently: looking up the delegate matching the candidate and updating supervisor-delegate records.

**Review by SteveJacksonSoft — CHANGES_REQUESTED** [review]

File: DigitalLearningSolutions.Data/Services/DelegateUploadFileService.cs
Comment (SteveJacksonSoft) [inline DigitalLearningSolutions.Data/Services/DelegateUploadFileService.cs:191]: Ah, good point - could we just tweak this to errorCodeOrCandidateNumber? That makes it a bit clearer that it's not a candidate number when an error's occurred. Currently it's not so obvious when it's a status and when it's a candidate number. No need for re-review. Unless it demands approval before letting you merge.

Comment (DanBloxham-sw) [reply DigitalLearningSolutions.Data/Services/DelegateUploadFileService.cs]: Changed to errorCodeOrCandidateNumber. No need for re-review, so I will merge it once the checks pass.

**Pull request merged**
